sklego/meta.py

```diff
@@ -100,7 +152,69 @@ def __remove_groups_from_x(self, X):
         else:
             return np.delete(X, self.groups, axis=1)
 
-    def fit(self, X, y):
+    def __fit_grouped_estimator(self, X, y, value_columns, group_columns):
+        group_indices = X.groupby(group_columns).indices
+
+        try:
+            grouped_estimations = {
+                group: clone(self.estimator).fit(X.loc[indices, value_columns], y.loc[indices])
+                for group, indices in group_indices.items()
+            }
+        except KeyError:
+            raise KeyError("The indices of y and X should match")
+
+        return grouped_estimations
+
+    def __check_shrinkage_func(self):
+        """Validate the shrinkage function"""
+        group_lengths = [10, 5, 2]
+        expected_shape = np.array(group_lengths).shape
+        try:
+            result = self.shrinkage_function_(group_lengths)
+            if not isinstance(result, np.ndarray):
+                raise ValueError(f"shrinkage_function({group_lengths}) should return an np.ndarray")
+            if result.shape != expected_shape:
+                raise ValueError(f"shrinkage_function({group_lengths}).shape should be {expected_shape}")
+        except Exception as e:
+            raise ValueError(f"Caught an exception while checking the shrinkage function: {str(e)}")
+
+    def __set_shrinkage_function(self):
```

Comment:
I think a Python mantra is to "fail fast". Maybe makes sense to do these checks during the init (if `self.shrinkage` is given).

Consider:

```
shrink_options = {
   "constant": constant_shrinkage,
   "relative": relative_shrinkage,
   "min_n_obs": min_n_obs_shrinkage,
}
if isnstance(self.shrinkage_function, str) and self.shrinkage_function not in shrink_options:
   raise ValueError('The specified ....') 

self.shrinkage_function = shrink_options.get(self.shrinkage_function, self.shrinkage_function)
```

Maybe add another check to see if `self.shrinkage_function` is callable when not a `str`

Reply:
Included this, good idea!

Reply:
I moved it to `__init__` but moved it back to `fit` on suggestion of Matthijs because that works better with `sklearn.clone` which is used later on.